libbeat/publisher/queue/diskqueue/queue.go

```diff
@@ -0,0 +1,177 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package diskqueue
+
+import (
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/feature"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/libbeat/paths"
+	"github.com/elastic/beats/v7/libbeat/publisher/queue"
+)
+
+// Settings contains the configuration fields to create a new disk queue
+// or open an existing one.
+type Settings struct {
+	// The path on disk of the queue's main metadata file.
+	// Paths for data segment files are created by appending ".{segmentIndex}" to
+	// this path.
+	// If blank, the default is "queue.dat" within the beat's data directory.
+	Path string
+
+	// The size in bytes of one data page in the on-disk buffer. To minimize
+	// data loss if there is an error, this should match the page size of the
+	// target filesystem.
+	PageSize uint32
```

Comment:
Are we missing a SegmentSize? Just from the types it is difficult to tell how things will exactly work. But I presume we create multiple live segment files, creating a new one once we roll over. This would mean that freeing space requires us to delete a segment file.

If segment files are append only, do we still need to split it into pages? Maybe we can reduce some user facing settings here.

Reply:
Yes, segment size and related details will be added later, none of the metadata structures are yet meant to be comprehensive, this is just a first pass filling out the framework and documenting the parts that are there.

Pages are used because the error recovery depends on it. See the "error recovery" and "further simplifications" section in the design draft -- the reason we can still get good recovery guarantees without a transactional model is because each page has a metadata header. This is especially important because otherwise a bad write could corrupt the frame header position and lose all subsequent data for the entire spool file. By arranging in pages we can restrict data loss to only the pages that were actually corrupted. (This design is "append only" from the logical user perspective but the raw file access won't be since the trailing frame needs to be rewritten on update.)